modules/fundamental/src/purl/service/mod.rs

```diff
@@ -230,41 +231,84 @@ impl PurlService {
         }
     }
 
-    pub async fn purl_by_purl<C: ConnectionTrait>(
+    #[instrument(skip(self, connection), err(level=tracing::Level::INFO))]
+    pub async fn fetch_purl_details<C: ConnectionTrait, I: AsRef<str> + Debug>(
         &self,
-        purl: &Purl,
-        deprecation: Deprecation,
+        identifiers: &[I],
+        deprecated: Deprecation,
         connection: &C,
-    ) -> Result<Option<PurlDetails>, Error> {
-        let canonical = CanonicalPurl::from(purl.clone());
-        match qualified_purl::Entity::find()
-            .filter(qualified_purl::Column::Purl.eq(canonical))
-            .one(connection)
+    ) -> Result<HashMap<String, PurlDetails>, Error> {
+        let (purls, uuids): (Vec<_>, Vec<_>) = identifiers
+            .iter()
+            .partition(|key| key.as_ref().starts_with("pkg:"));
+
+        let purls: Vec<_> = purls
+            .iter()
+            .map(|k| Purl::from_str(k.as_ref()).map_err(Error::Purl))
+            .collect::<Result<_, _>>()?;
+
+        let uuids: Vec<_> = uuids
+            .iter()
+            .map(|k| Uuid::from_str(k.as_ref()).map_err(Error::Uuid))
+            .collect::<Result<_, _>>()?;
+
+        let details = self
+            .purls_by_purl(&purls, deprecated, connection)
             .await?
-        {
-            Some(purl) => Ok(Some(
-                PurlDetails::from_entity(None, None, &purl, deprecation, connection).await?,
-            )),
-            None => Ok(None),
+            .into_iter()
+            .map(|detail| (detail.head.purl.to_string(), detail))
+            .chain(
+                self.purls_by_uuid(&uuids, deprecated, connection)
+                    .await?
+                    .into_iter()
+                    .map(|detail| (detail.head.uuid.to_string(), detail)),
+            )
+            .collect();
+
+        Ok(details)
+    }
+
+    async fn purls_by_purl<C: ConnectionTrait>(
+        &self,
+        purls: &[Purl],
+        deprecation: Deprecation,
+        connection: &C,
+    ) -> Result<Vec<PurlDetails>, Error> {
+        let canonical: Vec<CanonicalPurl> = purls
```

Comment:
```suggestion
    ) -> Result<Vec<PurlDetails>, Error> {
        if purls.is_empty() {
            return Ok(Default::default());
        }
        let canonical: Vec<CanonicalPurl> = purls
```
Should probably put in a short-circuit return for an empty list...

Reply:
Why is it better over iterating an empty Vector?

Reply:
It's just to prevent hitting the database with that `qualified_purl::Entity::find()` call on line 282. I think that's going to execute whether the list is empty or not, and our db is busy enough as it is. :smile:

Reply:
Makes sense. Change applied.